client/blocks/reader-post-card/featured-video.jsx

```diff
@@ -15,7 +15,7 @@ import FeaturedImage from './featured-image';
 class FeaturedVideo extends React.Component {
 
 	static propTypes = {
-		thumbnailUrl: React.PropTypes.string,
+		thumbnailUrlPromise: React.PropTypes.object,
```

Comment:
when reading through the purpose of this component @gwwar I can definitely see the value in _not_ pushing this through Redux

 - It's not data that we reasonably expect to use in different parts of the app
 - It's definitely nothing that we want or need to persist across sessions

that being said, it's _side-effecting_ 100% and including it in the component will only make it harder to test. passing in a promise as a prop seems like a reasonable compromise actually, and I've been trying to reconsider promise-returning actions lately though I am not convinced of their value. something like this feels like an exception.

another approach I could see being valid here is a dreaded singleton store. this could be very little more than a mapping between input video URLs and output video thumbnails. it could live in the same folder as this component, handle the fetching, and handle the mapping. with a very basic API we could hold down the complexity too. 

```js
// thumbnail-mapper.js
const lastFetches = {};
const listeners = {};
const mapping = {};

/**
 * Mutates the stored mapping between video URLs and
 * thumbnail URLs then notifies listening callbacks of the
 * change. The callbacks will receive a pair of values with
 * the requested video URL and the thumbnail mapping.
 * Including both values will provide additional safety on
 * the component side just in case the callback gets called
 * after the desired video URL has changed. It can verify
 * in its callback that the returned URL matches the
 * expected URL.
 */
const mapAndNotify = url => thumbnailUrl => {
	lastFetches[ url ] = Date.now();
	mapping[ url ] = thumbnailUrl;
	get( listeners, url, [] ).forEach( f => f( [ url, thumbnailUrl ] ) )
};

/**
 * Performs the stateful fetch for the thumbnails
 */
const fetchThumbnail = url =>
	fetch( … )
		.then( … )
		.then( mapAndNotify( url ) )
		.catch( … );

const lastFetch = url => ( Date.now() - get( lastFetches, url, -Infinity ) );
const requestMapping = url => lastFetch( url ) > MAX_DURATION && fetchThumbnail( url )

/**
 * Mutates the `listeners` to include or remove callback functions
 * which will get called when the mappings for their associated urls
 * change or get added.
 *
 * Note that this would be simpler if we called each callback on every
 * change, but by limited to URL and shifting some burden to the
 * calling components, we can prevent updating all of the preview
 * images at the same time and risk flickering for the user.
 */
const modifyListeners = ( op, url, f ) => Object.assign( listeners, { [ url ]: op( get( listeners, url, [] ), [ f ] ) } );
export const unsubscribe = partial( modifyListeners, difference ); // remove callback from listeners
export const subscribe = ( url, f ) => {
	requestMapping( url ); // make sure we have or fetch the mapping
	modifyListeners( union, url, f ); // add callback to listeners
}

// thumbnail-provider.jsx
import { subscribe, unsubscribe } from './thumbnail-mapping';

const ThumbnailProvider = Component => props => class extends Component {
	constructor( ...args ) {
		super( ...args );

		this.state = { thumbnailUrl: null };
		props.url && subscribe( props.url, this.storeMapping );
	}

	componentDidUpdate( prevProps ) {
		const { url: prevUrl } = prevProps;
		const { url: nextUrl } = this.props;

		unsubscribe( prevUrl, this.storeMapping );
		nextUrl && subscribe( nextUrl, this.storeMapping );
	}

	storeMapping = ( [ url, thumbnailUrl ] ) => this.props.url === url && this.setState( { thumbnailUrl } );

	render() {
		return <Component { ...props } videoThumbnail={ this.state.thumbnailUrl } />
	}
}

// video-thumbnail.jsx
export const VideoThumbnail = ( { thumbnailUrl } ) => <img src={ thumbnailUrl } />;

export default ThumbnailProvider( VideoThumbnail );
```

so maybe I too went crazy in posting code here but this is just one of other possible ways to accomplish this. I'm not sure how I feel about it in comparison to using Redux state and actions. I _do_ like the separation from persisted data and not; I _do_ like storing the temporary/ephemeral data in component state and keeping the cycle small. can't provide much more insight at this moment though.